File: python/source/interactive_books/domain/chunk_data.py
from dataclasses import dataclass

from interactive_books.domain.errors import BookError, BookErrorCode


@dataclass(frozen=True)
class ChunkData:
    content: str
    start_page: int
    end_page: int
    chunk_index: int

    def __post_init__(self) -> None:
        if not self.content:
            raise BookError(
                BookErrorCode.PARSE_FAILED,
                "ChunkData content cannot be empty",
            )
        if self.start_page < 1:
            raise BookError(
                BookErrorCode.PARSE_FAILED,
                f"ChunkData start_page must be >= 1, got {self.start_page}",
            )
        if self.end_page < self.start_page:
            raise BookError(
                BookErrorCode.PARSE_FAILED,
                f"ChunkData end_page ({self.end_page}) must be >= start_page ({self.start_page})",
            )
        if self.chunk_index < 0:
            raise BookError(
                BookErrorCode.PARSE_FAILED,
                f"ChunkData chunk_index must be >= 0, got {self.chunk_index}",
            )

File: swift/InteractiveBooks/Sources/InteractiveBooksCore/Domain/ValueObjects/ChunkData.swift
public struct ChunkData: Sendable, Equatable {
    public let content: String
    public let startPage: Int
    public let endPage: Int
    public let chunkIndex: Int

    public init(
        content: String,
        startPage: Int,
        endPage: Int,
        chunkIndex: Int
    ) throws {
        guard !content.isEmpty else {
            throw BookError.parseFailed("ChunkData content cannot be empty")
        }
        guard startPage >= 1 else {
            throw BookError.parseFailed(
                "ChunkData start_page must be >= 1, got \(startPage)"
            )
        }
        guard endPage >= startPage else {
            throw BookError.parseFailed(
                "ChunkData end_page (\(endPage)) must be >= start_page (\(startPage))"
            )
        }
        guard chunkIndex >= 0 else {
            throw BookError.parseFailed(
                "ChunkData chunk_index must be >= 0, got \(chunkIndex)"
            )
        }
        self.content = content
        self.startPage = startPage
        self.endPage = endPage
        self.chunkIndex = chunkIndex
    }
}
